File: code/ex3.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0, lower = -1;
        HashMap<Character, Integer> map = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            if (!map.containsKey(s.charAt(i))) {
                map.put(s.charAt(i), i);
                max = Math.max(max, i - Math.max(lower, -1));
            } else {
                max = Math.max(max, i - Math.max(map.get(s.charAt(i)), lower));
                // lower is to record the position of last duplicated char
                lower = Math.max(map.get(s.charAt(i)), lower);
                map.put(s.charAt(i), i);
            }
        }
        
        return max;
    }
}

File: code/ex3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        unordered_set<char> set;
        int MAX = 0, lo = 0, hi = 0;
        while (hi < n) {
            char chi = s[hi];
            while (lo < hi && set.find(chi) != set.end()) {
                set.erase(s[lo++]);
            }
            MAX = max(MAX, hi - lo + 1);
            set.insert(s[hi++]);
        }
        return MAX;
    }
};